text/0000-externally-implementable-functions.md

```diff
@@ -0,0 +1,147 @@
+- Feature Name: `extern_impl_fn`
+- Start Date: 2024-05-10
+- RFC PR: [rust-lang/rfcs#0000](https://github.com/rust-lang/rfcs/pull/0000)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+
+A mechanism for defining a function whose implementation can be defined (or overridden) in another crate.
+
+Example 1:
+
+```rust
+// core::panic:
+
+extern impl fn panic_handler(_: &PanicInfo) -> !;
+
+// user:
+
+impl fn core::panic::panic_handler(_: &PanicInfo) -> ! {
+    loop {}
+}
+```
+
+Example 2:
+
+```rust
+// log crate:
+
+extern impl fn logger() -> Logger {
+    Logger::default()
+}
+
+// user:
+
+impl fn log::logger() -> Logger {
+    Logger::to_stdout().with_colors()
+}
+```
+
+# Motivation
+
+We have several items in the standard library that are overridable/definable by the user crate.
+For example, the (no_std) `panic_handler`, the global allocator for `alloc`, and so on.
+
+Each of those is a special lang item with its own special handling.
+Having a general mechanism simplifies the language and makes this functionality available for other crates, and potentially for more use cases in core/alloc/std.
+
+# Explanation
+
+A function can be defined as "externally implementable" using `extern impl` as follows:
+
+```rust
+// In crate `x`:
+
+// Without a body:
+extern impl fn a();
+
+// With a body:
+extern impl fn b() {
+    println!("default impl");
+}
+```
+
+Another crate can then provide (or override) the implementation of these functions using `impl fn` syntax (using their path) as follows:
+
+```rust
+// In another crate:
+
+impl fn x::a() {
+    println!("my implementation of a");
+}
+
+impl fn x::b() {
+    println!("my implementation of b");
+}
+```
+
+# Details
```

Comment:
Is there a particular reason you diverged from the standard RFC template?

Reply:
This worked better. Otherwise I'l' just end up repeating the same thing twice. 🤷‍♀️

(I use the template as a useful guide, not as a fixed form to match. Most of my RFCs don't use the exact template structure.)

Reply:
We should have common guidance on how to use the template. I've (sometimes painfully) shoved my RFCs into the structure as I thought it was mandatory (and I was given that impression when I used an outdated version of the template) and now I see other people ignore the structure -- that's frustrating.

Reply:
Yeah I think it depends a bit on the situation how useful the template is. For complex language features, the template works well I think. But for smaller/simpler features or subtle changes it often doesn't work. For example, in [RFC 3007](https://rust-lang.github.io/rfcs/3007-panic-plan.html) I used a problems+solutions structure, which worked much better for a set of subtle changes that fix a set of subtle problems.

The risk of not following the RFC template is missing important parts, or producing a hard to follow document with unclear structure, or producing something that's not easily digestible by all target audiences. It's definitely a useful template.

But in the end the point of an RFC is to just clearly lay out all the relevant information. So I mostly see the template as a guide to cover all the information that should be in the document, not as a strict form that needs to be filled in in one exact way. (Especially if that'd just lead to a less clear presentation of the information, as that doesn't benefit anyone.)

Perhaps we should add a note to the template. Or provide multiple templates so one can pick whichever fits best.